crates/node/src/network_server/grpc_svc_handler.rs

```diff
@@ -282,6 +283,46 @@ impl NodeCtlSvc for NodeCtlSvcHandler {
 
         Ok(Response::new(cluster_state_response))
     }
+
+    async fn remove_node(
+        &self,
+        request: Request<RemoveNodeRequest>,
+    ) -> Result<Response<()>, Status> {
+        let request = request.into_inner();
+        let metadata = Metadata::current();
+        let node_id = PlainNodeId::from(request.node_id);
+
+        if metadata.nodes_config_ref().version() != Version::from(request.nodes_config_version) {
+            return Err(Status::failed_precondition(
+                "The nodes configuration version does not match the current version",
+            ));
+        };
+
+        self.metadata_store_client
+            .read_modify_write::<NodesConfiguration, _, _>(
+                NODES_CONFIG_KEY.clone(),
+                |nodes_config| {
+                    let mut nodes_config = nodes_config
+                        .ok_or(Status::not_found("Missing cluster nodes configuration"))?;
+
+                    if nodes_config.version() != Version::from(request.nodes_config_version) {
+                        return Err(Status::failed_precondition(
+                            "The nodes configuration version does not match the request version; try again",
+                        ));
+                    }
+
+                    nodes_config.remove_node(&node_id).map_err(|err| {
+                        Status::invalid_argument(format!("Failed to remove node: {err}"))
+                    })?;
+                    nodes_config.increment_version();
+
+                    Ok(nodes_config)
+                },
+            )
+            .await.map_err(|err| Status::internal(err.to_string()))?;
+
+        Ok(Response::new(()))
+    }
```

Comment:
Unfortunately, it's not that simple, but we can allow removing nodes if their log-server storage state is `Provisioning` or `Disabled`. For metadata-server this is still an open subject that's not fully designed yet.

Without managing cluster membership transition for the stateful roles, this can cause unrecoverable dataloss. Additionally, nodes that are alive might not know how to handle being deleted while running (unless this is something you have tested).

Reply:
Super useful, this is an additional level of safety I can add 🙏 I was hoping for some input from Till about metadata servers, but I don't want to distract right now.

Right now I'm only doing to safety check (is the node flagged as dead?) in client-side in `restatectl`, but the handler should be the actual gatekeeper.

Reply:
Ok, so this is potentially only useful for decommissioned admin | worker nodes in the mean time - and perhaps log servers with a bit of care. I'll keep the branch on my fork but close the PR for now to avoid polluting this space.